### PR TITLE
Brotli compression

### PR DESCRIPTION
**Pull request by trevyn**
Draft PR: Only the compression side of Brotli so far, it generates a file of roughly the correct size. I'll attack the decompression side later, after which we will see if the compression side *actually* works. ;)

### CODE REVIEW

**Comment by trevyn** [comment]
Ok, this actually works round-trip for me when rebuilding the stub by hand, but the `stubs` npm script fails with

`go build github.com/google/brotli/go/cbrotli: build constraints exclude all Go files in /Users/eden/go/pkg/mod/github.com/google/brotli/go/cbrotli@v0.0.0-20210127140805-63be8a994019`

I think this is because the Brotli library is actually in C, and CGO can't cross-compile. I'll take a closer look tomorrow.

**Comment by leafac** [comment]
I appreciate the work so far 😃

Keep going…

**Comment by leafac** [comment]
> @leafac I can think of two main options here:

Thanks for the investigation and for presenting the ideas so clearly 👍

I really like where this is going…

> * https://github.com/andybalholm/brotli is a translation of the reference Brotli implementation into pure Go, which should cross-compile without a problem. A disadvantage is that it is not the official implementation. Decompression speed would also need to be verified.
> * We could set up a system to use GitHub Actions to compile the stubs. Having 3x 3MB binaries checked into the git repo seems a little awkward to me, so this could also be a larger effort that would have the packager download the stubs from GitHub Releases or similar. This also has the advantage of providing a verified supply chain for the binaries, which is probably a good idea anyway.
> 
> Do you have a preference?

In principle I’d go with the reference implementation (cbrotli), because it seems to be the more principled approach. At this point you probably also know [my feelings about cross-compilation](https://github.com/leafac/caxa/tree/3097992a33d4e74b7ba99a940482e63d5484f7a6#why-no-cross-compilation-why-no-different-versions-of-nodejs-besides-the-version-with-which-caxa-was-called): Even if we were able to cross-compile, we’d still have to test our stuff on the target platform, so we might as well just compile there. (Ironically, I built the current stubs by cross-compiling, as you can tell from the `stubs` npm script in `package.json`. But I did my due diligence and tested the stubs myself in all platforms (and run the tests on GitHub Actions on them as well).)

Putting binaries in the Git repository is [how caxa currently works](https://github.com/leafac/caxa/tree/3097992a33d4e74b7ba99a940482e63d5484f7a6/stubs). It’s simple and effective, and I have nothing against it. Having the stubs in GitHub Releases and installing them on an `postinstall` npm script would also be fine, though it’s a bit more work.

One thing to keep in mind is that I don’t think building the stubs using GitHub Actions would be enough moving forward, as [there’s a demand for ARM support (Raspberry Pi, specifically)](https://github.com/leafac/caxa/issues/4) and as far as I can tell GitHub Actions doesn’t support that.

My first idea here is to get a Raspberry Pi, build the stub myself, test everything, and then distribute the stub.

But I also have another idea that I want to run by you: We could have a `postinstall` npm script that would download Go for your architecture and compile the stub for you. At this point, caxa becomes a weird kind of native module itself, I guess. Maybe we could provide pre-built stubs for the most popular platforms to speed things up for most people.

What do you think about all this?

**Comment by maxb2** [comment]
> One thing to keep in mind is that I don’t think building the stubs using GitHub Actions would be enough moving forward, as [there’s a demand for ARM support (Raspberry Pi, specifically)](https://github.com/leafac/caxa/issues/4) and as far as I can tell GitHub Actions doesn’t support that.

It is possible to emulate arm on Github actions using QEMU and docker buildx. Example [workflow](https://github.com/dungeon-revealer/dungeon-revealer/blob/ec85080cc4cfcd4e6a4b1247d9b72e9ff7abe99a/.github/workflows/build-binaries.yml#L60) and [Dockerfile](https://github.com/dungeon-revealer/dungeon-revealer/blob/ec85080cc4cfcd4e6a4b1247d9b72e9ff7abe99a/arm-compile.Dockerfile) to compile for armv7 and arm64.

**Comment by leafac** [comment]
Oh, that’s interesting. I was thinking about using QEMU for development until I got myself a Pi, but for some reason I never thought about doing it on GitHub Actions.

**Comment by leafac** [comment]
Do you strictly need Docker for that, or could you run QEMU directly on Ubuntu as provided by GitHub Actions? Isn’t it weird to run an emulator inside an already containerized environment?

**Comment by maxb2** [comment]
> Do you strictly need Docker for that

I don't think so, but it's really easy to set up this way. 

We use docker because it makes it really easy to pull build tools for a particular architecture. For example, there are node docker images [specifically made for armv7](https://hub.docker.com/layers/node/library/node/latest/images/sha256-a5d751d6009bf8efc7ae883ae969f063597ab1d52b1da9a3b16df248ee128768?context=explore). You can use `docker buildx build --platform linux/arm/v7 .` to automatically use the arm images. You can even use docker to setup qemu for buildx: `docker run --privileged --rm tonistiigi/binfmt --install arm64,arm`.

The basic idea is to create a Dockerfile that compiles whatever you want, use qemu+docker buildx to build the image, then run the image as a container and copy the files you want. 


In summary, with this example `Dockerfile` on your local machine:
```
FROM ubuntu

RUN your-compile-command outfile
```
You can run:
```
docker run --privileged --rm tonistiigi/binfmt --install arm
docker buildx build --platform linux/arm/v7 -t compiler-image .
APP=$(docker run --rm -it -d compiler-image:${{ matrix.arch }})
docker cp $APP:/path/to/outfile ./
docker kill $APP
```
Which will give you `outfile` in your current directory.

**Comment by leafac** [comment]
@maxb2 Thanks for the information. I played a little with Docker & ARM and I must admit it was a nice experience. I got more things working more quickly than when I tried to setup QEMU. In fact, all I did was:

```console
$ docker run -it arm64v8/ubuntu /bin/bash
```

And I was able to download both Node.js & Go for ARM8 and they ran just fine 🙌

Next I think I should learn more about buildx…

**Comment by leafac** [comment]
Okay, I think I learned all I need to know for now:

**Regarding my own question: “Isn’t it weird to run an emulator inside an already containerized environment?”:** That’s already how Docker works: https://www.docker.com/blog/multi-arch-images/ (see § How does it work?) In fact, as I mentioned above, running a Docker image for ARM was a breeze, while setting up QEMU by hand was a pain, so I can think of Docker as an easy-to-use emulator.

**Regarding buildx and the like:** That’s for when you want to build Docker images. We in caxa aren’t the business of building Docker images—we’re in the business of producing an npm package. In fact, not having to deal with Docker is a motivation to package a project using caxa in the first place. It’s amusingly ironic that Docker may help us get there. Same with Go, by the way: caxa exists so that I can have the Go deployment experience while continuing to write JavaScript—ironically, building the stub in Go was how we go there…

**In conclusion:** If we introduce CGO dependencies and lose the ability to cross-compile stubs, we can still use something like the `arm64v8/ubuntu` Docker image to build & test stuff (probably even on GitHub Actions?). We don’t need to deal with buildx for that. Also, we may continue to distribute the stubs in pre-compiled form: there are only a dozen platforms we’d have to support:

- Windows 32 bits.
- Windows 64 bits.
- macOS Intel.
- macOS ARM.
- Linux 64 bits.

**Comment by maxb2** [comment]
> **Regarding buildx and the like:** That’s for when you want to build Docker images. 

I was unaware that the ["u/arch"](https://github.com/docker-library/official-images#architectures-other-than-amd64) repos had such good coverage of useful tools. I was using buildx as a hack to access the other architectures in multi-arch manifests of official project images ([arm64v8/node](https://hub.docker.com/r/arm64v8/node) vs. [node with buildx](https://hub.docker.com/layers/node/library/node/latest/images/sha256-0aab0f340a0305d4bb68fe0e4c95772e3f42fb260a55f4ca0a2365eb67f351d8?context=explore)) So long as the `u/arch` repos are well maintained, that is a much better solution!

**Comment by leafac** [comment]
In the experiments above I was running Docker Desktop in macOS.

In any case, I’m happy with this setup. If you want to work on caxa for Linux/ARM you should:

- Ideally, get yourself a Linux/ARM setup, like a Raspberry Pi.
- Failing that, if you’re on Windows/macOS, use Docker Desktop.
- Failing that, use Docker buildx.

This list, of course, is ordered by complexity.

We can build the stubs and test everything using these rigs. If GitHub Actions works with the ARM Docker images, we may add then to the test matrix. If they don’t, we run the tests manually using the rigs mentioned above. We continue to distribute the stubs pre-compiled committed directly to the Git repository and included in npm package.

As much as I appreciate your work on https://github.com/leafac/caxa/pull/11 (I really do! 😃), I think that the solution is simpler and easier to maintain in the long run. Do you see any issues I may be missing?

**Comment by maxb2** [comment]
It is possible to use the `u/arch` images using standard docker if you [enable experimental features in the docker daemon](https://rolandsdev.blog/docker-experimental-in-github-workflow-for-ubuntu-18-04/) and use the `--platform` option. This eliminates the need for buildx. I've updated #11 to reflect this and it should be much more straight forward now.

As far as maintaining the stubs, I think it would be better to use a github workflow as it transparently shows the compilation process. As an extreme example, someone could open a PR with changes to `stub.go` and bad binaries. There isn't an easy way to verify that the binary is actually from the source file. There is a potential for bad actors or just headaches for maintainers debugging a bad binary. Having a transparent compilation process would increase trust in the binaries. In addition, compiling on actual hardware creates a barrier for those that want to contribute. Using a workflow to compile for all platforms allows developers to quickly test their changes rather than manually synchronizing and testing their changes on all the different machines required (we can very easily add a smoke test in the workflow).

You would have to change how you distribute the binaries (not having them committed in the repo itself), but I think the benefits outweigh the inconvenience. 

Sorry @trevyn for polluting your PR with all this docker talk :smile:

**Comment by leafac** [comment]
Some information that may be useful in this investigation about other compression solutions: https://github.com/leafac/caxa/issues/14#issue-843547055

**Comment by leafac** [comment]
Hi everyone,

First of all, thank you very much for your contribution! I appreciate the time you took into using caxa, understanding the packaging architecture, and contributing with the conversation about a different compression mechanism to make caxa better.

I’ve been thinking about the broad strategy employed by caxa and concluded that there is a better way to solve the problem. It doesn’t include a self-extracting executable, so it subsumes this pull request.

It’s a different enough approach that I think it deserves a new name, and it’s part of a bigger toolset that I’m building, which I call [**Radically Straightforward · Package**](https://github.com/radically-straightforward/radically-straightforward/tree/main/package).

I’m deprecating caxa and archiving this repository. I invite you to continue the conversation in [Radically Straightforward’s issues](https://github.com/radically-straightforward/radically-straightforward/issues).

Best.
